### PR TITLE
Add peopleHr connector implementation

### PR DESCRIPTION
**Pull request by SanduDS**
# Description
Providing a ballerina connector for [peopleHR](https://www.peoplehr.com/)

Fixes https://github.com/wso2-enterprise/choreo/issues/8243, https://github.com/ballerina-platform/ballerina-extended-library/issues/116

One line release note: 
- Ballerina PeopleHR connector implementation

## Type of change
- [x] New connector

# How Has This Been Tested?
 - Only operations related to Query
**Test Configuration**:
* Ballerina Version: Ballerina Swan Lake3
* Operating System: Ubuntu
* Java SDK: Java 11

# Checklist:

### Security checks
 - [x] Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? 
 - [x] Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets?

### CODE REVIEW

File: peoplehr/client.bal [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+        return self.basicClient->post(path, request, targetType = CheckAuthenticationResponse);
+    }
+
+    # Updates employee id details.
+    #
+    # + payload - Request details on an employee
+    # + return - Successful response or an error
+    @display{label: "Update EmployeeId"}   
+    isolated remote function updateEmployeeId(@display{label: "Request Detail"}UpdateEmployeeIDRequest payload) returns  UpdateEmployeeIdResponse|error {
+        string path = string `/Employee`;
```
Comment (LakshanSS) [inline peoplehr/client.bal]: This is repeated in many places. Shall we use a constant for this?

File: .github/workflows/daily-build.yml [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+              run: chmod +x gradlew
+
+            # Build the project with Gradle
+            - name: Build with Gradle
+              env:
+                packageUser: ${{ secrets.BALLERINA_BOT_USERNAME }}
+                packagePAT: ${{ secrets.BALLERINA_BOT_TOKEN }}
+                JAVA_OPTS: -DBALLERINA_DEV_COMPILE_BALLERINA_ORG=true
+              run: |
+                ./gradlew build
```
Comment (LakshanSS) [inline .github/workflows/daily-build.yml]: Shall we remove the gradle related steps here?
(If we don't have any java component, they are not required)

File: .github/workflows/daily-build.yml [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+              run: |
+                curl \
+                -X POST 'https://chat.googleapis.com/v1/spaces/${{secrets.BALLERINA_CHAT_ID}}/messages?key=${{secrets.BALLERINA_CHAT_KEY}}&token=${{secrets.BALLERINA_CHAT_TOKEN}}' \
+                --header 'Content-Type: application/json' \
+                -d '{"text": "*module-ballerinax-peoplehr* daily build failure \nPlease visit <https://github.com/ballerina-platform/module-ballerinax-peoplehr/actions?query=workflow%3A%22Daily+build%22|the daily build page> for more information"}'
+
+                curl \
+                -X POST 'https://chat.googleapis.com/v1/spaces/${{secrets.CONNECTOR_CHAT_ID}}/messages?key=${{secrets.CONNECTOR_CHAT_KEY}}&token=${{secrets.CONNECTOR_CHAT_TOKEN}}' \
+                --header 'Content-Type: application/json' \
+                -d '{"text": "*module-ballerinax-peoplehr* daily build failure \nPlease visit <https://github.com/ballerina-platform/module-ballerinax-peoplehr/actions?query=workflow%3A%22Daily+build%22|the daily build page> for more information"}'
```
Comment (LakshanSS) [inline .github/workflows/daily-build.yml]: Now we dont send notification to 2 groups.
We send notification only to one common group.
Please check the implementation in other repos.

File: gradle.properties [hunk 4]
```diff
@@ -0,0 +1,4 @@
+org.gradle.caching=true
+group=org.ballerinalang.peoplehr
+version=0.1.0-SNAPSHOT
```
Comment (LakshanSS) [inline gradle.properties]: Version should be 1.0.0
Lets not add SNAPSHOT

File: peoplehr/Ballerina.toml [hunk 5]
```diff
@@ -0,0 +1,12 @@
+[package]
+org = "ballerinax"
+name = "peoplehr"
+version = "0.1.0"
```
Comment (LakshanSS) [inline peoplehr/Ballerina.toml]: Lets change the version as 1.0.0

File: peoplehr/Ballerina.toml [hunk 6]
```diff
@@ -0,0 +1,12 @@
...
+distribution = "slbeta3"
+authors = ["Ballerina"]
+repository = "https://github.com/ballerina-platform/module-ballerinax-peoplehr"
+keywords = ["Cost/Paid"]
```
Comment (LakshanSS) [inline peoplehr/Ballerina.toml]: need to add more keywords according to the categorization

File: peoplehr/client.bal [hunk 7]
```diff
@@ -0,0 +1,265 @@
...
+        request.setJsonPayload(check jsonBody.mergeJson(authentication));
+        return check self.basicClient->post(path, request, targetType = GetVacancyResponse);
+    }
+
+    # Gets all vacancy detail
+    #
+    # + return - Successful response or an error
+    @display{label: "Get Vacancies"}
+    isolated remote function getAllVacancies() returns GetAllVacanciesResponse|error {
+        string path = string `/Vacancy`;
```
Comment (LakshanSS) [inline peoplehr/client.bal]: Used in many places.
Can use a constant for this

File: peoplehr/types.bal [hunk 8]
```diff
@@ -0,0 +1,856 @@
...
+# Represents common fields of the response
+#
+# + isError - Indicates whether the response is an error or not
+# + Status - The status of the error  
+# + Message - The message of the response  
+public type CommonResponse record {
+    boolean isError;
+    int Status;
+    string Message;
+
```
Comment (LakshanSS) [inline peoplehr/types.bal]: Shall we remove unnecessary new line here?

File: peoplehr/Package.md [hunk 9]
```diff
@@ -0,0 +1,18 @@
...
+This package provides access to operations to mange to employee details, holidays and salary in your company.
+
+### Compatibility
+|                      | Version                   |
+|----------------------|---------------------------|
+| Ballerina Language   | Ballerina Swan Lake Beta3 |
+| People HR API        | 3.1                       |
+
+## Report issues
+To report bugs, request new features, start new discussions, view project boards, etc., go to the [Ballerina connector repository](https://github.com/ballerina-platform/module-ballerina-peoplehr)
```
Comment (LakshanSS) [inline peoplehr/Package.md]: This has to point to the extended-library repo.
Please check in other connector repos

File: README.md [hunk 10]
```diff
@@ -1,2 +1,47 @@
-# module-ballerinax-peoplehr
 Ballerina connector for People HR
+===================
+[![CI](https://github.com/SanduDS/module-ballerinax-peoplehr/actions/workflows/ci.yml/badge.svg)](https://github.com/SanduDS/module-ballerinax-peoplehr/actions/workflows/ci.yml)
```
Comment (LakshanSS) [inline README.md]: Shall we add the correct link here?

File: peoplehr/.settings/org.eclipse.buildship.core.prefs [hunk 11]
```diff
@@ -0,0 +1,2 @@
+connection.project.dir=..
+eclipse.preferences.version=1
```
Comment (SkNuwanTissera) [inline peoplehr/.settings/org.eclipse.buildship.core.prefs]: Why do we use this file?

File: peoplehr/client.bal [hunk 12]
```diff
@@ -0,0 +1,260 @@
...
+        request.setJsonPayload(check jsonBody.mergeJson(authentication));
+        return check self.basicClient->post(VACANCY, request, targetType = AddNewVacancyResponse);
+    }
+
+}
+
+# Configuration record for PeopleHR.
+#
+# + apiKey - PeopleHR API key  
+# + baseURL - Bas URL
```
Comment (SkNuwanTissera) [inline peoplehr/client.bal]: `# + baseURL - Base URL`

File: peoplehr/types.bal [hunk 13]
```diff
@@ -0,0 +1,854 @@
...
+# Represents the response type of `updateEmployeeDetail` operation  
+public type UpdateEmployeeDetailResponse record {|
+    *UpdateEmployeeIdResponse;
+|};
+
+# Represents the response type of `updateEmployeeDetail` operation
+#
+# + Result - The response details of the operation 
+public type CheckAuthenticationResponse record {|
+    *CommonResponse;
```
Comment (abeykoon) [inline peoplehr/types.bal]: Shall we rename to `PeopleHrResponse`?

**Comment by abeykoon** [comment]
We decided to perform this change. 

When we get defined fields and some more, we can keep the record as closed but open to `json`. That's the best experience we can give via Ballerina lang. When we access, we can use following notation for additional fields 

```
string value = record.["myKey"];
```

File: peoplehr/Ballerina.toml [hunk 14]
```diff
@@ -0,0 +1,13 @@
+[package]
```
Comment (LakshanSS) [inline peoplehr/Ballerina.toml:1]: need to add distribution (slbeta6)

File: peoplehr/Dependencies.toml [hunk 15]
```diff
@@ -0,0 +1,302 @@
+# AUTO-GENERATED FILE. DO NOT MODIFY.
```
Comment (LakshanSS) [inline peoplehr/Dependencies.toml:1]: Need to delete this and regenerate with SLBeta6

File: peoplehr/Package.md [hunk 16]
```diff
@@ -0,0 +1,18 @@
+Connects to [PeopleHR API](https://apidocs.peoplehr.com/#section/Introduction) from Ballerina
+
+## Package overview
+The peoplehr is a [Ballerina](https://ballerina.io/) connector for People HR API.
+This package provides access to operations to mange to employee details, holidays and salary in your company.
+
+### Compatibility
+|                      | Version                   |
+|----------------------|---------------------------|
+| Ballerina Language   | Ballerina Swan Lake Beta3 |
```
Comment (LakshanSS) [inline peoplehr/Package.md]: Need to update to Beta6

File: README.md [hunk 17]
```diff
@@ -1,2 +1,47 @@
...
+ 
+For more information, see [Ballerina Documentation - ballerinax/peoplehr](https://central.ballerina.io/ballerinax/peoplehr/latest).
+ 
+## Building from the source
+### Setting up the prerequisites
+1. Download and install Java SE Development Kit (JDK) version 11. You can install either [OpenJDK](https://adoptopenjdk.net/) or [Oracle JDK](https://www.oracle.com/java/technologies/javase-jdk11-downloads.html).
+ 
+       > **Note:** Set the JAVA_HOME environment variable to the path name of the directory into which you installed JDK.
+ 
+2. Download and install [Ballerina Swan Lake Beta3](https://ballerina.io/)
```
Comment (LakshanSS) [inline README.md]:
```suggestion
2. Download and install [Ballerina Swan Lake](https://ballerina.io/)
```

File: gradle.properties [hunk 18]
```diff
@@ -0,0 +1,4 @@
+org.gradle.caching=true
+group=org.ballerinalang.peoplehr
+version=1.0.0
+ballerinaLangVersion=2.0.0-beta.3-20210821-093200-e064dcef
```
Comment (LakshanSS) [inline gradle.properties]: We can remove this since this connector doesn't have a java wrapper

File: peoplehr/client.bal [hunk 19]
```diff
@@ -0,0 +1,260 @@
...
+// specific language governing permissions and limitations
+// under the License.
+
+import ballerina/http;
+
+# This connector helps you easily integrate People with other systems and applications, for seamless cross-platform data sharing. The People API 
+# accepts and returns JSON data in the request body, with status indicating the outcome of the operation (sucess/failure).
+#
+# + basicClient - People HTTP Client
+@display {label: "PeopleHR Client", iconPath: "resources/peoplehr.svg"}
```
Comment (abeykoon) [inline peoplehr/client.bal]: We need to update the icon path.

File: peoplehr/client.bal [hunk 20]
```diff
@@ -0,0 +1,260 @@
...
+public isolated client class Client {
+    final http:Client basicClient;
```
Comment (abeykoon) [inline peoplehr/client.bal]: basicClient --> HTTPClient?

File: peoplehr/client.bal [hunk 21]
```diff
@@ -0,0 +1,260 @@
...
+        self.config = config.cloneReadOnly();
+        self.basicClient = check new (config.baseURL, httpClientConfig);
+    }
+
+    # Checks authentication for given user.
+    #
+    # + payload - Request details on authentication
+    # + return -  Successful response or an error
+    @display{label: "Check Authentication"} 
+    isolated remote function checkAuthentication(@display{label: "Authentication Detail"} AuthenticationRequest payload)
```
Comment (abeykoon) [inline peoplehr/client.bal]: what does this method actually do? Is it to check if a given user is authenticated at peopleHR? What is returned in the response?

File: peoplehr/client.bal [hunk 22]
```diff
@@ -0,0 +1,260 @@
...
+        request.setJsonPayload(check jsonBody.mergeJson(authentication)); 
+        return self.basicClient->post(EMPLOYEE, request, targetType = UpdateEmployeeIdResponse);
+    }
+
+    # Gets employee detail list.
+    #
+    # + payload - Request details on all employees
+    # + return - Successful response or an error
+    @display{label: "Get All Employee Detail"}
+    isolated remote function getAllEmployeeDetail(@display{label: "Request Detail"}AllEmployeeDataRequest payload)
```
Comment (abeykoon) [inline peoplehr/client.bal]: Shall we rename to `getAllEmployees`?

File: peoplehr/client.bal [hunk 23]
```diff
@@ -0,0 +1,260 @@
...
+        request.setJsonPayload(check jsonBody.mergeJson(authentication));
+        return self.basicClient->post(EMPLOYEE, request, targetType = GetAllEmployeeDetailResponse);
+    }
+
+    # Gets employee detail by id
+    #
+    # + payload - Request details on an employee
+    # + return - Successful response or an error
+    @display{label: "Get Employee Detail By Id"}
+    isolated remote function getEmployeeDetailById(@display{label: "Employee Request Detail"} EmployeeDetailRequest
```
Comment (abeykoon) [inline peoplehr/client.bal]: Shall we rename to `getEmployeeById`?

File: peoplehr/client.bal [hunk 24]
```diff
@@ -0,0 +1,260 @@
...
+    @display{label: "Update Employee Detail"}   
+    isolated remote function updateEmployeeDetail(@display{label: "Employee Detail"} EmployeeUpdateRequest|json payload) 
+                                                  returns UpdateEmployeeDetailResponse|error {
+        http:Request request = new;
+        json jsonBody = check payload.cloneWithType(json);
+        json authentication =  {APIKey: self.config.apiKey, Action: "UpdateEmployeeDetail"}; 
+        request.setJsonPayload(check jsonBody.mergeJson(authentication));
+        return self.basicClient->post(EMPLOYEE, request, targetType = UpdateEmployeeDetailResponse);
+    }
+
```
Comment (abeykoon) [inline peoplehr/client.bal:110]: Shall we reorganize the remote methods to be in CRUD order? That's the order the drop down at Choreo would list down the operations. If you see update operation in the dropdown first, it is kind of weird.

File: peoplehr/client.bal [hunk 25]
```diff
@@ -0,0 +1,260 @@
...
+        request.setJsonPayload(check jsonBody.mergeJson(authentication)); 
+        return self.basicClient->post(HOLIDAY, request, targetType = DeleteHolidayResponse);      
+    }
+
+    # Gets query result By query name details
+    #
+    # + payload - Request detail on a query 
+    # + return - Successful response or an error
+    @display{label: "Get Query Result By Query Name"}
+    isolated remote function getQueryResultByQueryName(@display{label: "Request Detail on Query Result"}
```
Comment (abeykoon) [inline peoplehr/client.bal]: I am not clean what this method does. Is it `getQueryByName`? PeopleHR users would know there will be something called queries.

File: peoplehr/types.bal [hunk 26]
```diff
@@ -0,0 +1,863 @@
...
+    string APIKey?;
+    string Action?;
+};
+
+# Represents common fields of the response
+#
+# + isError - Indicates whether the response is an error or not
+# + Status - The status of the error  
+# + Message - The message of the response  
+public type PeopleHrResponse record {
```
Comment (abeykoon) [inline peoplehr/types.bal]: Shall we rename to `PeopleHRGenricResponse`? Check if we can make this a private type as users will not use it directly.

File: peoplehr/client.bal [hunk 27]
```diff
@@ -0,0 +1,260 @@
...
+    # + config - People connection configuration
+    # + httpClientConfig - HTTP configuration
+    # + return - `http:Error` in case of failure to initialize or `null` if successfully initialized 
+    public isolated function init(ConnectionConfig config, http:ClientConfiguration httpClientConfig = {}) returns
+                                error? {
+        self.config = config.cloneReadOnly();
+        self.basicClient = check new (config.baseURL, httpClientConfig);
+    }
+
+    # Checks authentication for given user.
```
Comment (abeykoon) [inline peoplehr/client.bal]: Check if credentails of a given login user is valid.

File: peoplehr/types.bal [hunk 28]
```diff
@@ -0,0 +1,863 @@
...
+# + Result - The response details of the operation 
+public type CheckAuthenticationResponse record {|
+    *PeopleHrResponse;
+    json Result;
+|};
+
+# Represents the response type of `UpdateEmployeeId` operation
+#
+# + Result - The response of the operation 
+public type UpdateEmployeeIdResponse record {|
```
Comment (abeykoon) [inline peoplehr/types.bal]: Shall we rename to `OperationStatus` and remove individual records which includes this alone?

File: peoplehr/client.bal [hunk 29]
```diff
@@ -0,0 +1,261 @@
...
+        request.setJsonPayload(authentication);
+        return check self.HTTPClient->post(VACANCY, request, targetType = AllVacancies);
+    }
+
+    # Gets query result By query name details.
+    #
+    # + payload - Request detail on a query 
+    # + return - Successful response or an error
+    @display{label: "Get Query Result By Query Name"}
+    isolated remote function getQueryByName(@display{label: "Request Detail on Query Result"}
```
Comment (abeykoon) [inline peoplehr/client.bal:225]: Shall we rename
`GetQueryByNameResponse ` --> `QueryDetail`?

**Review by abeykoon — APPROVED** [review]

**Pull request merged**